python/ray/autoscaler/_private/kuberay/node_provider.py

```diff
@@ -0,0 +1,295 @@
+import json
+import logging
+import requests
+import threading
+import time
+from typing import Any, Dict, List, Tuple
+
+from ray.autoscaler.node_provider import NodeProvider
+from ray.autoscaler.tags import (STATUS_UP_TO_DATE, STATUS_UPDATE_FAILED)
+
+# Terminology:
+
+# Labels and Tags
+# We call the Kuberay labels "labels" and the Ray autoscaler tags "tags".
+# The labels are prefixed by "ray.io". Tags are prefixed by "ray-".
+# We convert between the two but do not mix them.
+
+# Worker Groups and Available Node Types
+# In Kuberay the different node types are called "worker groups", in the
+# the Ray autoscaler they are called "available node types".
+
+# Design:
+
+# Each modification the autoscaler wants to make is posted to the API server goal state
+# (e.g. if the autoscaler wants to scale up, it increases the number of
+# replicas of the worker group it wants to scale, if it wants to scale down
+# it decreases the number of replicas and adds the exact pods that should be
+# terminated to the scaleStrategy). In order to guarantee consistency, the NodeProvider
+# then waits until Kuberay's reconciliation loop creates the pod specifications in the
+# API server and then returns control back to the autoscaler. The waiting period
+# is typically small, on the order of a few seconds. We make sure that only one
+# such modification is in process by serializing all modification operations with
+# a lock in the NodeProvider.
+
+# Note: Log handlers set up in autoscaling monitor entrypoint.
+logger = logging.getLogger(__name__)
+
+provider_exists = False
+
+
+def to_label_selector(tags: Dict[str, str]) -> str:
+    "Convert tags to label selector to embed in query to K8s API server."
+    label_selector = ""
+    for k, v in tags.items():
+        if label_selector != "":
+            label_selector += ","
+        label_selector += "{}={}".format(k, v)
+    return label_selector
+
+
+def status_tag(pod: Dict[str, Any]) -> str:
+    """Convert pod state to Ray autoscaler status tag."""
+    if "containerStatuses" not in pod["status"]:
```

Comment:
nit, optional: add `or not pod["status"]["containerStatuses"]` to the condition just for good measure